[JENKINS-55448] Util delete regression

jvz opened the pull request:
See [JENKINS-55448](https://issues.jenkins-ci.org/browse/JENKINS-55448). This is a followup to https://github.com/jenkinsci/jenkins/pull/3812

<!-- Comment: 
If the issue is not fully described in the ticket, add more information here (justification, pull request links, etc.).

 * We do not require JIRA issues for minor improvements.
 * Bugfixes should have a JIRA issue (backporting process).
 * Major new features should have a JIRA issue reference.
-->

### Proposed changelog entries

* Fix `jenkins.util.io.CompositeIOException` thrown where workspaces and other directories/files could not be deleted when any of its ancestor files are a symbolic link or junction point or contain path entries of `.` or `..`. ([issue 55448](https://issues.jenkins-ci.org/browse/JENKINS-55448), [issue 55450](https://issues.jenkins-ci.org/browse/JENKINS-55450))
* Internal: Add path checking support to `jenkins.util.io.PathRemover` to help simplify `hudson.FilePath` deletion methods.
* Internal: Revert behavior of `hudson.Util.isSymlink()` to previous algorithm.

<!-- Comment: 
The changelogs will be integrated by the core maintainers after the merge.  See the changelog examples here: https://jenkins.io/changelog/ -->

### Submitter checklist

- [x] JIRA issue is well described
- [x] Changelog entry appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
      * Use the `Internal: ` prefix if the change has no user-visible impact (API, test frameworks, etc.)
- [x] Appropriate autotests or explanation to why this change has no tests
~~- [ ] For dependency updates: links to external changelogs and, if possible, full diffs~~

<!-- For new API and extension points: Link to the reference implementation in open-source (or example in Javadoc) -->

### Desired reviewers

@reviewbybees @jeffret-b @daniel-beck @kevinearls @MRamonLeon @fcojfernandez 
@jenkinsci/code-reviewers

jvz committed:
Add test for deleting a symlink

Signed-off-by: Matt Sicker <boards@gmail.com>

jvz committed:
Update FilePath.delete* to use Util.delete*

This reduces duplication of file deletion strategy code and potentially fixes a regression noted in JENKINS-55448.

Signed-off-by: Matt Sicker <boards@gmail.com>

jvz committed:
Add test for deleting symlinks in descendant files

Signed-off-by: Matt Sicker <boards@gmail.com>

jvz committed:
[JENKINS-55448] Fix Util.delete regression

This adds path normalization behavior back to the delete methods to match the behavior of the previous version of these methods.

Signed-off-by: Matt Sicker <boards@gmail.com>

dwnusbaum commented on core/src/main/java/hudson/FilePath.java:
```diff
@@ -1271,7 +1271,7 @@ public void deleteRecursive() throws IOException, InterruptedException {
         @Override
         public Void invoke(File f, VirtualChannel channel) throws IOException {
-            deleteRecursive(deleting(f));
+            Util.deleteRecursive(deleting(f));
```
The previous implementation of `deleteRecursive` would call `deleting(...)` on each file in the directory. Does the new `Util.deleteRecursive` still do this? If not, I think this might be a problem for agent-master security.

jvz replied:
Oh, great point! I'll fix this, thanks.

jvz committed:
Add path filter support to PathRemover

This allows for simplifying FilePath to use PathRemover rather than duplicating logic.

Signed-off-by: Matt Sicker <boards@gmail.com>

dwnusbaum commented on core/src/main/java/hudson/FilePath.java:
```diff
@@ -1271,7 +1271,7 @@ public void deleteRecursive() throws IOException, InterruptedException {
         @Override
         public Void invoke(File f, VirtualChannel channel) throws IOException {
-            deleteRecursive(deleting(f));
+            Util.deleteRecursive(fileToPath(f), path -> deleting(path.toFile()).delete());
```
I don't think you want to call `File.delete` in the filter, since it will attempt to delete the file, and if it fails, you'll won't get an exception, since it is an old IO, not an NIO call, right? You probably want something more like: 

```
path -> {
try { 
    deleting(path.toFile())
} catch (Exception e) { // Or maybe specifically SecurityException
    return false; // Better to propagate errors though...
}
return true; // If allowed, let `Files.deleteIfExists` handle it for better error messages in case of failure.
```

That said, in the interest of propagating the actual error to users, I would suggest using a functional interface with a method like `void check(Path p) throws SecurityException` instead of `Predicate<Path>`.

jvz replied:
Yeah, I was considering making a functional interface like that.

And good catch. I mixed up the delete() that performs the security check with the delete() that actually deletes the file.

jvz committed:
Revert Util.isSymlink behavior

The simplified form that this replaces has an issue where a file contained somewhere inside a symbolic link is also considered a symbolic link on Windows.

Signed-off-by: Matt Sicker <boards@gmail.com>

jvz committed:
Use explicit functional interface for path checking

Signed-off-by: Matt Sicker <boards@gmail.com>

jeffret-b reviewed (approved):
Looks like it should work.

fcojfernandez reviewed (approved):
Code LGTM. Are the failures in tests related to this change? Have you analysed?

batmat commented:
+1 with Francisco, failures _look_ unrelated, but I think we must get an assessment from your side @jvz
![image](https://user-images.githubusercontent.com/223853/51033389-8105bc00-15a3-11e9-9067-a2f881a07523.png)

batmat reviewed (approved):
Thanks a lot for working on improving the situation Matt!

I tried your change locally, and after reverting your changes under core/src/main/java, running the tests does fail, which is what I expected to approve. Great job :+1:!

```
$ git fetch origin pull/3841/head:PR-3841-jvz-util-delete-regression
$ git co PR-3841-jvz-util-delete-regression
$ git co 4c6160d710f903884d6268caed14794bccb5372a~ core/src/main/java
$ mvn clean -pl core -am test -Dtest='Path*Test,Util*Test'
[... snip ...]
[INFO] 
[INFO] Results:
[INFO] 
[ERROR] Errors: 
[ERROR]   PathRemoverTest.testForceRemoveRecursive_ContainsDotPath:364 » CompositeIO Una...
[ERROR]   PathRemoverTest.testForceRemoveRecursive_ParentIsSymbolicLink:382 » CompositeIO
[INFO] 
[ERROR] Tests run: 34, Failures: 0, Errors: 2, Skipped: 5
[INFO] 
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary for Jenkins main module 2.158-SNAPSHOT:
[INFO] 
[INFO] Jenkins main module ................................ SUCCESS [  1.648 s]
[INFO] Jenkins cli ........................................ SUCCESS [  4.252 s]
[INFO] Jenkins core ....................................... FAILURE [ 32.602 s]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  39.664 s
[INFO] Finished at: 2019-01-11T13:22:12+01:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:2.20:test (default-test) on project jenkins-core: There are test failures.
[ERROR]
[ERROR] Please refer to /home/tiste/dev/github/jenkinsci/jenkins/core/target/surefire-reports for the individual test results.
[ERROR] Please refer to dump files (if any exist) [date]-jvmRun[N].dump, [date].dumpstream and [date]-jvmRun[N].dumpstream.
[ERROR] -> [Help 1]
[ERROR]
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR]
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException
[ERROR]
[ERROR] After correcting the problems, you can resume the build with the command
[ERROR]   mvn <goals> -rf :jenkins-core
```

jvz committed:
Add more tests for JENKINS-55448

This also refactors some test calls to mklink into a package-private Util method due to additional tests needing to create junction points.

Signed-off-by: Matt Sicker <boards@gmail.com>

jvz commented:
I added a few more tests to exercise things related to the regression. I also added one more `Util` method, but this one is for test usage only, so I made it package-private. If anyone knows a more appropriate location for this code, please let me know!

fcojfernandez commented on core/src/main/java/hudson/Util.java:
```diff
@@ -1244,6 +1278,38 @@ public static String resolveSymlink(@Nonnull File link) throws IOException {
+     * Escapes a Windows cmd argument. Cmd supports escaping characters with carats which simplifies nested escapes.
+     * @param arg argument to escape
+     * @return the quoted argument for use in cmd.exe
+     * @see <a href="https://blogs.msdn.microsoft.com/twistylittlepassagesallalike/2011/04/23/everyone-quotes-command-line-arguments-the-wrong-way/>Everyone quotes command line arguments the wrong way</a>
```
The href attribute is not properly closed
```suggestion
     * @see <a href="https://blogs.msdn.microsoft.com/twistylittlepassagesallalike/2011/04/23/everyone-quotes-command-line-arguments-the-wrong-way">Everyone quotes command line arguments the wrong way</a>
```

fcojfernandez committed:
Fix typo in link

Co-Authored-By: jvz <boards@gmail.com>

jeffret-b reviewed (approved):
It still looks good to me. The new tests are a good addition.

I'll plan on merging after the build / tests complete so this regression fix can go in the next release.

jvz commented:
> It would probably help to adjust changelog entries to look something like these: https://jenkins.io/changelog/ . Specifically anything that says "bug" should reference the Jira ticket. I think the best proposed changelog entry proposal would be a relatively short single entry. Then move the additional details about the several specific issues into the Jira ticket.

Changelog updated to be more in line with the style of the official changelog.

jvz committed:
Fix Windows command line argument quoting in createJunction

Signed-off-by: Matt Sicker <boards@gmail.com>

jvz committed:
Move Windows util code into test sources

Signed-off-by: Matt Sicker <boards@gmail.com>

jvz committed:
Add link to blog post

Signed-off-by: Matt Sicker <boards@gmail.com>

jvz committed:
Improve cmd quoting algorithm

Signed-off-by: Matt Sicker <boards@gmail.com>

jtnord commented on core/src/main/java/hudson/Util.java:
```diff
@@ -1290,26 +1289,16 @@ public static String resolveSymlink(@Nonnull File link) throws IOException {
+                WindowsCommandLineFormatter.quoteArgumentForCmd(junction.getAbsolutePath()) +
+                ' ' +
+                WindowsCommandLineFormatter.quoteArgumentForCmd(target.getAbsolutePath());
+        Process process = Runtime.getRuntime().exec(command);
```
would this have been easier if you used the array version rather than the single string version?

jtnord commented on core/src/test/java/hudson/os/WindowsUtil.java:
```diff
@@ -0,0 +1,117 @@
+     */
+    public static @Nonnull Process execCmd(String... argv) throws IOException {
+        String command = Arrays.stream(argv).map(WindowsUtil::quoteArgumentForCmd).collect(Collectors.joining(" "));
+        return Runtime.getRuntime().exec(new String[]{"cmd.exe", "/C", command});
```
this smells bad - you are passing in an array but then collapsing it to a single argument for the call to exex?  why not just keep the array - is there a reason (does cmd.exe loose that information)?

jvz replied:
The problem is that without escaping, there is no way to distinguish between whitespace in a file name and whitespace denoting different command line args. If we passed all the args properly to cmd.exe, cmd.exe will lose the quoting when it evals the command given. Windows is more annoying than bash in nested escaping.

jvz replied:
Heh, that link is in the code comments for this class.

jvz committed:
Fix test failure

Signed-off-by: Matt Sicker <boards@gmail.com>

jvz committed:
Merge branch 'master' into util-delete-regression

Signed-off-by: Matt Sicker <boards@gmail.com>

jvz commented:
Due to the revert of some related changes, the current diff is now essentially a combination of the previous PR and the fixes to them. In either case, you can review this holistically (including the prior changes for reducing test failures in UtilTest), or as just the original changes. I found the test failure that prevented this from being merged last week (I had refactored a test incorrectly) and am now waiting for it to pass all tests. Otherwise, this can likely be merged for the next rolling release.

jeffret-b commented:
It still looks good. This is the right way to go. Too bad we couldn't get this correction in the latest release, but now we've got a little more time to get it in.

jeffret-b commented:
Removed the `regression` label as since the original change was reverted this doesn't really qualify as a regression any more. Now this is more the correct fix to the original problem.
